src/generic/gridsel.cpp

```diff
@@ -398,517 +288,262 @@ void wxGridSelection::SelectBlock( int topRow, int leftCol,
-                    n--;
-                    count--;
-                    break;
-
-                default:
-                    break;
-            }
-        }
-
-        // If a row containing the selection is already selected, return,
-        // if a row contained in newly selected block is found, remove it.
-        count = m_rowSelection.GetCount();
-        for ( n = 0; n < count; n++ )
-        {
-            switch ( BlockContain( m_rowSelection[n], 0,
-                                   m_rowSelection[n], m_grid->GetNumberCols() - 1,
-                                   topRow, leftCol, bottomRow, rightCol ) )
-            {
-                case 1:
-                    return;
-
-                case -1:
-                    m_rowSelection.RemoveAt(n);
-                    n--;
-                    count--;
-                    break;
-
-                default:
-                    break;
-            }
-        }
-
-        // Same for columns.
-        count = m_colSelection.GetCount();
-        for ( n = 0; n < count; n++ )
-        {
-            switch ( BlockContain( 0, m_colSelection[n],
-                                   m_grid->GetNumberRows() - 1, m_colSelection[n],
-                                   topRow, leftCol, bottomRow, rightCol ) )
-            {
-                case 1:
-                    return;
-
-                case -1:
-                    m_colSelection.RemoveAt(n);
-                    n--;
-                    count--;
-                    break;
-
-                default:
-                    break;
-            }
-        }
-
-        m_blockSelectionTopLeft.Add( wxGridCellCoords( topRow, leftCol ) );
-        m_blockSelectionBottomRight.Add( wxGridCellCoords( bottomRow, rightCol ) );
-    }
-    // Update View:
-    if ( !m_grid->GetBatchCount() )
-    {
-        m_grid->RefreshBlock(topRow, leftCol, bottomRow, rightCol);
-    }
+    EnsureFirstLessThanSecond(topRow, bottomRow);
+    EnsureFirstLessThanSecond(leftCol, rightCol);
 
-    // Send Event, if not disabled.
-    if ( sendEvent )
-    {
-        wxGridRangeSelectEvent gridEvt( m_grid->GetId(),
-            wxEVT_GRID_RANGE_SELECT,
-            m_grid,
-            wxGridCellCoords( topRow, leftCol ),
-            wxGridCellCoords( bottomRow, rightCol ),
-            true,
-            kbd);
-        m_grid->GetEventHandler()->ProcessEvent( gridEvt );
-    }
+    Select(topRow, leftCol, bottomRow, rightCol, kbd, sendEvent);
 }
 
 void wxGridSelection::SelectCell( int row, int col,
                                   const wxKeyboardState& kbd,
                                   bool sendEvent )
 {
-    if ( IsInSelection ( row, col ) )
+    int topRow = row;
+    int leftCol = col;
+    int bottomRow = row;
+    int rightCol = col;
+
+    // Fix the coordinates if needed.
+    switch (m_selectionMode)
+    {
+    default:
+        wxFAIL_MSG("unknown selection mode");
+        wxFALLTHROUGH;
+
```

Comment:
It would be best to avoid having `default` here, as this inhibits warnings about not handling some enum elements. The best would be to have something like this:
```
int allowed = -1;
switch (m_selectionMode) {
    case SelectCells | SelectRows | SelectColumns: ...; allowed = 1; break;
    case SelectRowsOrColumns: allowed = 0; break;
}

wxASSERT_MSG(allowed != -1, "unknown selection mode?" );
if ( !allowed )
     return;
```